README.md
[![Build Status](https://travis-ci.org/matutter/dockerfilejs.svg?branch=master)](https://travis-ci.org/matutter/dockerfilejs) [![Coverage Status](https://coveralls.io/repos/github/matutter/dockerfilejs/badge.svg?branch=master)](https://coveralls.io/github/matutter/dockerfilejs?branch=master)

# Dockerfilejs
#### A simple Dockerfile generator. [Try it now!](https://tonicdev.com/npm/dockerfilejs)
#### [Based on the official Dockerfile Reference](https://docs.docker.com/engine/reference/builder/)

```javascript
var Dockerfile = require("dockerfilejs").Dockerfile;
var file = new Dockerfile();

file.comment('The above code example yields this file!')
  .env({DEBUG:'express:* node index.js'})
  .expose(8080)
  .separator('\n')
  .from({ image : 'node', tag : 'latest'})
  .comment('FROM gets bumped under initial comments')
  .render();
```

```Dockerfile
# The above code example yields this file!
FROM node:latest
ENV DEBUG="express:* node index.js"
EXPOSE 8080
# FROM gets bumped under initial comments
```

#### Complex environment? No problem!
```javascript
file.env({
    complex: {
        objects:  'are not',
        a: 'problem',
        at : {
            all: 'really!'
        }
}}).render();
/*
ENV complex.objects="are not" \
    complex.a=problem \
    complex.at.all=really!
*/
```

#### Made a mistake? Get rid of it!
```javascript
file.copy('~/.ssh/*', '/tmp');
// oops!
file.steps().pop();
```

#### Advanced examples
```javascript
file.separator('\n')
// Will set the separator of each step for the entire file.render() output

file.label({ complex: { objects: 'allowed' } });
// LABEL complex.objects="allowed"

file.expose([8080, '8081', { number: 443, protocol: 'tcp' }]);
// EXPOSE 8080 8081 443/tcp

file.run({ command: ['touch /file.txt', ['echo', 'hello world', '>>', '/file.txt'] ] });
// RUN touch /file.txt \
//   && echo "hello world" >> /file.txt

file.copy({ src : ['/id_rsa', '/id_rsa.pub'], dest: '/root/.ssh/' }, true);
// ONBUILD COPY ["/id_rsa", "/id_rsa.pub", "/root/.ssh"]

file.copy({ src : ['/id_rsa', '/id_rsa.pub'], dest: '/root/.ssh/', onbuild: true });
// ONBUILD COPY ["/id_rsa", "/id_rsa.pub", "/root/.ssh"]

file.copy({ src : ['/id_rsa'], dest: '/root/.ssh/', chown: '0:0' });
// COPY --chown=0:0 ["/id_rsa", "/root/.ssh"]

file.copy({ src : ['/id_rsa'], dest: '/root/.ssh/', user: 0, group: 0 });
// COPY --chown=0:0 ["/id_rsa", "/root/.ssh"]

file.copy({ src : ['/id_rsa'], dest: '/root/.ssh/', chown: { user: 0, group: 0 });
// COPY --chown=0:0 ["/id_rsa", "/root/.ssh"]

file.cmd({ executable: '/bin/bash', params: ['-c', 'hello world'] });
// CMD ["/bin/bash", "-c", "hello world"]

file.cmd({ command:'/bin/bash', params: ['-c', 'hello world'] })
// CMD /bin/bash -c "hello world"

file.healthCheck({
    options: { retries: 4, timeout: '30s'},
    command: 'wget',
    params: ['example.com']
})
// HEALTHCHECK --retries=4 --timeout=30s \
//   CMD wget example.com

file.user('root');
// USER root
```

#### Multi-stage builds
```javascript
file.from({ image: 'node', tag: '10-alpine', name: 'build' })
// FROM node:10-alpine AS build

// ... run your build commands here ...

file.stage()
// Adds a new stage

file.from({ image: 'node', registry: 'docker.io', tag: '10-alpine' })
// Sets the `FROM` instruction in the new stage, etc ...

file.copy({ from: 'build', src: ['./lib'], dest: './lib' })
// COPY --from=build ["./lib", "./lib"]
```


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

## [v1.2.1] 2019-07-20
### Modified
- lodash dependency v4.17.11 upgreaded to 4.17.15


test/test.js
module.exports = require('..')



test/testInstructions.js
var ins = require('..').instructions
var assert = require('chai').assert

const command = {
  "ADD" : ins.add,
  "COPY" : ins.copy,
  "ENV" : ins.env,
  "EXPOSE" : ins.expose,
  "LABEL" : ins.label,
  "USER" : ins.user,
  "WORKDIR" : ins.workdir,
  "VOLUME" : ins.volume, 
  "STOPSIGNAL" : ins.stopSignal 
}

var instr_exports = [
  "from",       "maintainer", "run",          "cmd",
  "entryPoint", "label",      "expose",       "env",
  "add",        "copy",       "user",         "workdir",
  "stopSignal", "volume",     "healthCheck",  "shell",
  "comment"
]

var onBuild_methods = [
  "ADD",  "COPY",  "ENV",  "EXPOSE", 
  "LABEL",  "USER",  "WORKDIR",  "VOLUME", 
  "STOPSIGNAL"
]

var valid_inputs = {
  ADD : {src: '/file.txt', dest: '/file.txt'},
  COPY : {src: '/file.txt', dest: '/file.txt'},
  ENV : {key: 'val', key2: 'val 2'},
  EXPOSE : [8000, '8001', {number: 433, protocol: 'tcp'}, {from: 8080, to:8086}],
  LABEL : { key: 'val', nested: {key: 'val', doubley: { key: 'val'}}},
  USER : 'user1',
  WORKDIR : '/git/',
  VOLUME : ['/var/log'],
  STOPSIGNAL : 9,
}

describe('"instructions" are exported from the main module', ()=> {
  it('should export each instruction from https://docs.docker.com/engine/reference/builder/', ()=> {
    var ins_exports_found = Object.keys(ins)
    assert.deepEqual(instr_exports, ins_exports_found)
  })
})

describe('All onBuild instruction are passed true as a second parameter', ()=> {
  it('should prefix each with ONBUILD', ()=> {
    onBuild_methods.forEach(method_name => {
      var command_name = method_name.toUpperCase()
      var input = valid_inputs[command_name]
      var res = command[command_name](input, true)
      assert(res.startsWith('ONBUILD'))
    })
  })
})

/*

var res = null 

res = ins.from({image: "alpine"})
res = ins.run({command: ['touch /file.txt', ['echo', 'hello world', '>>', '/file.txt']]})
res = ins.run({command: ['echo', 'hello world', '>>', '/file.txt']})
res = ins.from({image: "alpine", registry: 'my_registry.io'})
//res = ins.cmd({executable: '/bin/bash', command: 'echo'})
res = ins.cmd({executable: '/bin/bash', params: ['-c', 'echo', 'hello world', '>>', '/file.txt']})
res = ins.cmd({executable: '/bin/bash'})
res = ins.cmd({command: 'touch /file.txt'})
res = ins.cmd({params: ['hello world', '>>', '/file.txt']})
//res = ins.cmd({params: null})
res = ins.label({ key: 'val', nested: {key: 'val', doubley: { key: 'val'}}}, true)
//res = ins.label({ })
res = ins.entryPoint(['/bin/bash', '-c'])
//res = ins.entryPoint(null)
//res = ins.entryPoint('')
//res = ins.entryPoint([])
res = ins.expose([8000, '8001', {number: 433, protocol: 'tcp'}, {from: 8080, to:8086}])
//res = ins.expose([{from: 8080}])
res = ins.expose([{number: 433}], true)
res = ins.env({key: 'val', key2: 'val 2'})
//res = ins.add({src: '', dest: '/file.txt'})
res = ins.add({src: '/file.txt', dest: '/file.txt'})
res = ins.copy({src: '/file.txt', dest: '/file.txt'})
res = ins.volume(['/var/log', '/root/.ssh'])
res = ins.healthCheck({
  options : {
    interval: '5m',
    timeout: '3s',
    retries: 5
  },
  command : 'curl',
  params: ['-f', 'http://localhost/', '||', 'exit', 1]
})
res = ins.healthCheck(false)
res = ins.comment('I am a', 100, 'year old' , 'Wizard')
res = ins.shell(['cmd', '/S', '/C'])



console.log(res)
*/

test/instruction/test_add.js
var ins = require('../test.js').instruction
var assert = require('chai').assert

describe('testing instruction add', ()=> {

  describe('with a valid string for the src and dest properties', ()=> {
    const expect = 'ADD ["src", "dest"]'
    it('should produce ' + expect, ()=> {
      const result = ins.add({src: 'src', dest: 'dest'})
      assert.equal(result, expect)
    })
  })

  describe('with an array of sources', ()=> {
    const expect = 'ADD ["src1", "src2", "src3", "dest"]'
    it('should produce ' + expect, ()=> {
      const result = ins.add({src: ['src1', 'src2', 'src3'], dest: 'dest'})
      assert.equal(result, expect)
    })
  })

  describe('with the onbuild parameter set to true run(..., true)', ()=> {
    const expect = 'ONBUILD ADD ["src1", "src2", "src3", "dest"]'
    it('should produce '+expect, ()=> {
      const result = ins.add({src: ['src1', 'src2', 'src3'], dest: 'dest'}, true)
      assert.equal(result, expect)
    })
  })

  describe('with chown flag (shorthand)', ()=> {
    const expect = 'ADD --chown=42 ["src", "dest"]'
    it('should produce '+expect, ()=> {
      const result = ins.add({ src: ['src'], dest: 'dest', chown: 42 })
      assert.equal(result, expect)
    })
  })

  describe('with chown flag (longhand)', ()=> {
    const expect = 'ADD --chown=user:group ["src", "dest"]'
    it('should produce '+expect, ()=> {
      const result = ins.add({ src: ['src'], dest: 'dest', chown: 'user:group' })
      assert.equal(result, expect)
    })
  })

  describe('for invalid input', ()=> {
    it('should throw on either null', ()=> {
      assert.throws(() => { ins.add({src: null, dest:null}) })
      assert.throws(() => { ins.add({src: 'src', dest:null}) })
      assert.throws(() => { ins.add({src: null, dest:'dest'}) })
    })

    it('should throw on either empty', ()=> {
      assert.throws(() => { ins.add({src: '', dest:''}) })
      assert.throws(() => { ins.add({src: '', dest:'dest'}) })
      assert.throws(() => { ins.add({src: 'src', dest:''}) })
    })

    it('should throw on either empty array', ()=> {
      assert.throws(() => { ins.add({src: [], dest:'dest'}) })
      assert.throws(() => { ins.add({src: ['asd'], dest:[]}) })
    })
  })

})

lib/model.js
const DockerfileModel = {
  from : {
    image : null,
    tag : null,
    digest : null,
    registry : null,
    stage : null
  },
  maintainer : {
    name : null
  },
  run : {
    command : [null]
  },
  cmd : {
    command : null,
    executable : null,
    params : [null]
  },
  entryPoint : {
    executable: null,
    params: null
  },
  label: {},
  expose : [
    '',
    0,
    {
      number: 0,
      protocol: ''
    }
  ],
  env : {},
  add : {
    src: [],
    dest: '',
    chown: ''
  },
  copy: {
    src: [],
    dest: '',
    chown: '',
    from: ''
  },
  user: '',
  workdir: '',
  stopSignal: '',
  volume : [''],
  healthCheck : {
    options: {
      interval : '',
      timeout: '',
      retries: 0
    },
    command : null,
    executable : null,
    params : [null]
  },
  shell : [],
  comment: ''
}

module.exports.Dockerfile = DockerfileModel

dockerfile.js
const instruction = require('./lib/instruction.js')
const model = require('./lib/model.js')

function InstructionWrap(name, data, on_build_flag, fn) {
  this.name = name
  this.data = data
  this.on_build_flag = on_build_flag
  this.fn = fn
}

function Dockerfile() {
  var steps = []
  this.stages = [this]
  this._current_stage = 0
  this._separator = '\n\n'
  this.steps = () => steps
  this.append = (step) => {
    this.getStage().steps().push(step);
    return this
  }
  this.prepend = (step) => {
    this.getStage().steps().unshift(step);
    return this
  }
  this.splice = (start, step) => {
    this.getStage().steps().splice(start, 0, step)
    return this
  }
}
Object.keys(instruction).forEach(name => {
  const fn = instruction[name]
  if(name == 'from') {
    Dockerfile.prototype[name] = function(data, on_build_flag) {
      var from = new InstructionWrap(name, data, on_build_flag, fn)
      var start = this.steps().findIndex(step => step.name != 'comment')
      this.splice(start, from)
      return this
    }
  } else {
    Dockerfile.prototype[name] = function(data, on_build_flag) {
      this.append(new InstructionWrap(name, data, on_build_flag, fn))
      return this
    }
  }
})

/** set step / section separator */
Dockerfile.prototype.separator = function(separator) {
  this.stages.forEach(stage => stage._separator = separator);
  return this
};

Dockerfile.prototype.renderSteps = function() {
  // unused parameter
  return this
    .steps()
    .map(step => step.fn(step.data, step.on_build_flag).trim())
};

Dockerfile.prototype.render = function(steps) {
  return this.stages.map(stage => {
    return stage.renderSteps(steps).join(this._separator || '\n')
  }).join(this._separator || '\n')
};

Dockerfile.prototype.stage = function(stage) {
  if (stage) {
    // Rely on index error ...
    var _ = this.stages[stage];
  } else {
    var df = new Dockerfile();
    // TODO: this is fine for now - there are few things to copy
    df.separator(this._separator);
    stage = this.stages.push(df) - 1;
  }

  // _current_stage == null is the same as make "this" one active
  this._current_stage = stage;

  return this;
}

Dockerfile.prototype.getStage = function() {
  return this.stages[this._current_stage];
}

module.exports.Dockerfile = Dockerfile
module.exports.instruction = instruction
module.exports.model = model


lib/instruction-error.js
'use strict';

module.exports.InstructionError = InstructionError;
module.exports.MissingPropertyError = MissingPropertyError;
module.exports.EmptyArrayError = EmptyArrayError;
module.exports.EmptyStringError = EmptyStringError;
module.exports.PropertyConflictError = PropertyConflictError;

const util = require('util');
const fmt = util.format;
const inherits = util.inherits;

function InstructionError(instruction, message) {
  Error.captureStackTrace(this, this.constructor);
  this.name = this.constructor.name;
  this.instruction = instruction.toUpperCase();
  this.message = fmt('Instruction "%s" %s', this.instruction, message);
};
inherits(InstructionError, Error)

function MissingPropertyError(instruction, property, found) {
  Error.captureStackTrace(this, this.constructor);
  this.name = this.constructor.name;
  this.instruction = instruction.toUpperCase();
  this.message = fmt('Instruction "%s" expected property "%s", found %s', this.instruction, property, found);
};
inherits(MissingPropertyError, Error)

function EmptyArrayError(instruction, property) {
  Error.captureStackTrace(this, this.constructor);
  this.name = this.constructor.name;
  this.instruction = instruction.toUpperCase();
  this.message = fmt('Instruction "%s" expected array property "%s" of length 1 or more', this.instruction, property);
};
inherits(EmptyArrayError, Error)

function EmptyStringError(instruction, property) {
  Error.captureStackTrace(this, this.constructor);
  this.name = this.constructor.name;
  this.instruction = instruction.toUpperCase();
  this.message = fmt('Instruction "%s" expected property "%s", found [EMPTY STRING]', this.instruction, property);
};
inherits(EmptyStringError, Error)

function PropertyConflictError(instruction, props) {
  Error.captureStackTrace(this, this.constructor);
  this.name = this.constructor.name;
  this.instruction = instruction.toUpperCase();
  this.message = fmt('Instruction "%s" encountered conflicting properties: %s', this.instruction, props.join(', '));
};
inherits(PropertyConflictError, Error)


lib/instruction.js
const errors = require('./instruction-error.js')

function boxAndFilter(arrayOrString) {
  if(!Array.isArray(arrayOrString))
    arrayOrString = [arrayOrString]

  return arrayOrString
    .filter(string => string) // not null or undefined
    .filter(string => string != '') // not empty
}

function onBuild(instruction) {
  return `ONBUILD ${instruction}`
}

/**
* Case in v18.09:
*   FROM < image > [AS < name > ]
*   FROM < image > [: < tag > ][AS < name > ]
*   FROM < image > [@ < digest > ][AS < name > ]
* Parameters:
*   image, (optional) tag, digest, registry
*
* Not standard:
*   If registry defined image is formatted as <registry>/<image>
*/
function from(from) {
  var instruction = null
  var image = from.image
  const tag = from.tag
  const digest = from.digest
  const registry = from.registry
  // The dockerfile reference referes to the AS argument as "name"
  // We'll except `stage` to be kind
  const name = from.name || from.stage

  if(!image && typeof from === 'string') {
    image = from
  }

  if(image && (image = image.trim()) && image.length) {
    if(registry) {
      image = registry+'/'+image
    }
    if(tag) {
      instruction = `FROM ${image}:${tag}`
    } else if(digest) {
      instruction = `FROM ${image}@${digest}`
    } else {
      instruction = `FROM ${image}`
    }
  } else {
    throw new errors.MissingPropertyError('from', 'image', image)
  }

  if (name) {
    instruction += ` AS ${name}`
  }

  return instruction
}

function maintainer(maintainer) {
  var instruction = null
  const name = maintainer.name || maintainer

  if(name && typeof name === 'string') {
    instruction = `MAINTAINER ${name}`
  } else {
    throw new errors.MissingPropertyError('maintainer', 'name', name)
  }

  return instruction
}

function enQuote(string) {
  return '"'+string+'"'
}

function mapToQuote(string) {
  if(/\s/g.test(string))
    string = enQuote(string)
  return string
}

/**
* Input string: returns trimmed string
* Input array of strings: returns a array of trimmed strings
*
* Note: Strings with WS in an array will be encapsulated by quotes
*/
function mapCommandParams(command) {
  if(Array.isArray(command)) {
    command = command.map(mapToQuote).join(' ')
  }
  return command.trim()
}

/**
* Case:
*   RUN <command>
* Parameters:
*   command (alias: commands)
*
* Expect string of well-formed command or an array of strings of
* well formed command and or arrays of command arguments.
* Example 1: well-formed command
*   run({command: 'touch /file.txt'})
* Example 2: array of well formed commands
*   run({command: ['touch /file.txt', ...]})
* Example 3: array of well formed commands and array of command argument
*   run({command: ['touch /file.txt', ['echo', 'hello world', '>>', '/file.txt']]})
*
* Note: Will trim leading and trailing WS
*/
function run(run) {
  var instruction = null
  var commands = run.command || run.commands || run

  if(commands == '') {
    throw new errors.EmptyStringError('cmd', 'command(s)')
  }

  if(commands) {
    if(typeof commands === 'string') {
      instruction = `RUN ${commands}`
    } else {
      commands = boxAndFilter(commands)

      if(commands.length) {
        instruction = `RUN ${commands.map(mapCommandParams).join(' \\\n  && ')}`
      } else {
        throw new errors.EmptyArrayError('cmd', 'command(s)')
      }
    }
  } else {
    throw new errors.MissingPropertyError('cmd', 'command(s)', commands)
  }

  return instruction
}

/**
* Case:
*   CMD ["executable","param1","param2"]
*   CMD ["param1","param2"]
*   CMD command param1 param2
*
* Parameters:
*   params and or (executable or command)
*
*   Params may be empty if and only if executable or command is provided
*
*   Solely providing the params property will yield the second case.
*   Providing the executable or command property will yield the first or third cases respectively.
*   Providing command and executable will throw an error
*/
function cmd(cmd) {
  var instruction = null
  var params = cmd.params
  const command = cmd.command
  const executable = cmd.executable

  if(command && executable) {
    throw new errors.PropertyConflictError('cmd', ['command', 'executable'])
  }

  if(params) {
    params = boxAndFilter(params)
  } else {
    params = null
  }

  if(executable) {
    if(params)
      instruction = `CMD ["${executable}", ${params.map(enQuote).join(', ')}]`
    else
      instruction = `CMD ["${executable}"]`
  } else if(command) {
    if(params)
      instruction = `CMD ${command} ${mapCommandParams(params)}`
    else
      instruction = `CMD ${command}`
  } else {
    if(params && params.length) {
      instruction = `CMD [${params.map(enQuote).join(', ')}]`
    } else {
      throw new errors.MissingPropertyError('cmd', 'params', params)
    }
  }


  return instruction
}

/**
* Case:
*   ENTRYPOINT ["executable", "param1", "param2"]
*
* Docs:
*   https://docs.docker.com/engine/reference/builder/#/entrypoint
*
* As the documentation points out, the "exec" form is preferred.
*
*/
function entryPoint(entrypoint) {
  var instruction = null

  const executable = entrypoint.executable
  entrypoint = boxAndFilter(entrypoint.params)

  if(entrypoint.length) {
    instruction = `ENTRYPOINT ["${executable}", ${entrypoint.map(mapToQuote).join(', ')}]`
  } else {
    throw new errors.InstructionError('entrypoint', 'expected string or array of arguments.')
  }

  return instruction
}

/**
* Flattens objects into an array of arrays, returns null if object has no OwnProperties.
*/
function flattenLabels(o, namespace) {
  namespace = namespace || []
  var r = []
  var ns = namespace.join('.')

  if(ns.length) ns = ns + '.'

  Object.keys(o).forEach( key => {
    const val = o[key]
    if(typeof val === 'string')
      return r.push([ns+key, val])

    r = r.concat(flattenLabels(val, namespace.concat([key])))
  })

  return r
}

/** Maps an array of ['key', 'val'] into a string of form 'key="val"' */
function mapLabel(label) {
  return label[0] + "=" + mapToQuote(label[1])
}

/**
* Case:
*   LABEL multi.label1="value1" multi.label2="value2" other="value3"
*
*
*/
function label(labels, onbuild) {
  labels = flattenLabels(labels)
  var instruction = null

  if(labels && labels.length) {
    instruction = `LABEL ${labels.map(mapLabel).join(' \\\n      ')}`
  } else {
    throw new errors.InstructionError('label', 'Provided object with no OwnProperties.')
  }

  return onbuild ?  onBuild(instruction) : instruction
}

/**
* Case:
*   EXPOSE <port> [<port>...]
*
* Expects an array of ports as numbers, strings, or an object of {number: <num or str>, protocol: <string> }
* If a port object is provided and the protocol field is blank it will be omitted.
*
* Example "port" object, assuming a EOLN standard object:
* var config = {
*   ports: [
*     {number: 443, protocol: 'tcp'},
*     8000
*   ]
* }
* instructions.expose(config.ports)
*
* A "port object" may also specify from and to for a range of exposed ports, though
* simply supplying '8080-8086' will do the same.
*
* Examples 1, shows all possible input type variants:
*   port(['1', 2, {number: 3, protocol: 'tcp'}, {number: '4', protocol: 'udp'}])
*/
function expose(expose, onbuild) {
  var instruction = null
  const trim = (s) => s.toString().trim()
  expose = boxAndFilter(expose)
    .map(port => {
      if(port.from && port.to) {
        port.number = `${trim(port.from)}-${trim(port.to)}`
      }
      if(port.number) {
        if(port.protocol) {
          return `${trim(port.number)}/${trim(port.protocol)}`
        } else {
          return trim(port.number)
        }
      } else if(typeof port === 'string' || typeof port === 'number') {
        return trim(port)
      } else {
        throw new TypeError('Expected object of type String, Number, or Port. Found:'+port)
      }
    })

  if(expose.length) {
    instruction = `EXPOSE ${expose.join(' ')}`
  } else {
    throw new errors.InstructionError(
      'expose',
      'expected one of string, number, Port object, [string or number or Port object, ...]'
    )
  }

  return onbuild ?  onBuild(instruction) : instruction
}

/**
* ENV <key> <value>
* ENV <key>=<value> ...
*
* Parameters:
*   1) Any object with terminal property-values of string, or number type.
*/
function env(env, onbuild) {
  var instruction = null

  env = flattenLabels(env)

  if(env.length) {
    instruction = `ENV ${env.map(e => `${e[0]}=${mapToQuote(e[1])}`).join(' \\\n    ')}`
  } else {
    throw new errors.InstructionError('env', 'expected string or array of arguments.')
  }

  return onbuild ?  onBuild(instruction) : instruction
}

// used to generate add and copy since they are of the same form
function copyInternal(command, src, dest, flags, onbuild) {
  var instruction = null
  src = boxAndFilter(src)

  if(src.length == 0) {
    throw new errors.EmptyArrayError(command, 'src')
  }

  flags.forEach(flag => {
    command += ` --${flag[0]}=${flag[1]}`
  })

  if(dest && dest.length) {
    instruction = `${command} [${src.map(enQuote).join(', ')}, "${dest}"]`
  } else {
    throw new errors.MissingPropertyError(command, 'dest')
  }

  return onbuild ?  onBuild(instruction) : instruction
}

/**
 * We'll take chown flags in a few ways. 
 * 1. An object with a `user` and/or `group` key.
 * 2. A `chown` property with a `user` and/or `group` key.
 * 
 * @returns [string] - May return an empty list.
 */
function getChownFlags(opts) {
  // Copy so we can mutate
  opts = Object.assign({}, opts)

  // special case - 0 is the only valid falsy value here
  if ( opts.chown === 0 ) opts.chown = '0'
  if ( opts.user === 0 ) opts.user = '0'
  if ( opts.group === 0 ) opts.group = '0'

  // any of these will produce some kind of --chown flag
  if ( !(opts.chown || opts.user || opts.group) ) {
    return null
  }

  let chown = opts.chown || { user: opts.user, group: opts.group };
  if ( chown.user || chown.group ) {
    // Its an object, force it to a string like user:group
    chown = `${chown.user || ''}:${chown.group || ''}`
    chown = chown.replace(/(^:)|(:$)/, '')
  } else {
    // Its a string or number so force it to string.
    chown = `${chown}`
  }

  if (chown.length) {
    return ['chown', chown]
  }

  return null
}

/**
* Usage:
*   add({src: '', dest: ''}, true)
*   add({src: '', dest: '', onbuild: true})
*   add({src: '', dest: ''})
*   add({src: '', dest: '', user: '', group: ''})
* This version overrides previous user & group
*   add({src: '', dest: '', chown: 'user:group'})
* This version overrides previous chown
*   add({src: '', dest: '', chown: { user: '', group: ''}})
*
* Case (* preferred):
*   ADD [--chown = < user >: < group > ] < src > ... < dest >
*   ADD [--chown = < user >: < group > ]["<src>", ..."<dest>"]
*/
function add(add, onbuild) {
  var instruction = null
  const src = add.src
  const dest = add.dest
  const flags = []

  let chown = getChownFlags(add);
  if (chown) {
    flags.push(chown);
  }

  // Copy this from the `add` object, makes easier use from serialized forms
  if ( onbuild == undefined && add.onbuild != undefined ) {
    onbuild = add.onbuild;
  }

  instruction = copyInternal('ADD', src, dest, flags, onbuild)
  return instruction
}

/**
* Case (* preferred):
*   COPY <src>... <dest>
*   * COPY ["<src>",... "<dest>"]
*/
function copy(copy, onbuild) {
  var instruction = null
  const src = copy.src
  const dest = copy.dest
  const flags = [];
  
  let chown = getChownFlags(copy);
  if (chown) {
    flags.push(chown);
  }

  if (copy.from) {
    flags.push(['from', copy.from])
  }

  if (onbuild == undefined && copy.onbuild != undefined) {
    onbuild = copy.onbuild;
  }

  instruction = copyInternal('COPY', src, dest, flags, onbuild)
  return instruction
}

function user(user, onbuild) {
  user = 'USER ' + user
  return onbuild ? onBuild(user) : user
}

function workdir(workdir, onbuild) {
  workdir = 'WORKDIR ' + workdir
  return onbuild ? onBuild(workdir) : workdir
}

function stopSignal(stopSignal, onbuild) {
  stopSignal = 'STOPSIGNAL ' + stopSignal
  return onbuild ? onBuild(stopSignal) : stopSignal
}

function volume(volume, onbuild) {
  volume = `VOLUME  [${volume.map(enQuote).join(', ')}]`
  return onbuild ? onBuild(volume) : volume
}

const valid_healthcheck_options = ['interval', 'timeout', 'retries']
/**
* Case:
*   HEALTHCHECK [OPTIONS] CMD command
*   HEALTHCHECK NONE
*
* Case 1: Pass an object with the properties for the CMD instruction, (optional) options
* Case 2: healthCheck(false)
*
* Docs: https://docs.docker.com/engine/reference/builder/#/healthcheck
*/
function healthCheck(healthCheck) {
  var instruction = null

  if(healthCheck == false) {
    instruction = 'HEALTHCHECK NONE'
  } else {
    var options = healthCheck.options

    if(options) {
      var keys = Object.keys(options)
        .filter(option => ~valid_healthcheck_options.indexOf(option))

      if(keys.length) {
        options = keys.map(k => {
          return `--${k}=${options[k]}`
        }).reduce((a,b) => a+' '+b)
      }
    }

    var command = cmd(healthCheck)

    if(options.length)
      instruction = `HEALTHCHECK ${options} \\\n  ${command}`
    else
      instruction = `HEALTHCHECK ${command}`
  }

  return instruction
}

/**
* Case:
*   SHELL ["executable", "parameters"]
*/
function shell(shell) {
  return `SHELL [${shell.map(enQuote).join(', ')}]`
}

function comment() {
  if(arguments.length == 0)
    return '#'
  var args = []
  for(var i = 0; i < arguments.length; ++i)
    args.push(arguments[i])
  return '# ' + args.map(arg => arg ? arg.toString() : arg).join(' ')
}

module.exports.from = from
module.exports.maintainer = maintainer
module.exports.run = run
module.exports.cmd = cmd
module.exports.entryPoint = entryPoint
module.exports.label = label
module.exports.expose = expose
module.exports.env = env
module.exports.add = add
module.exports.copy = copy
module.exports.user = user
module.exports.workdir = workdir
module.exports.stopSignal = stopSignal
module.exports.volume = volume
module.exports.healthCheck = healthCheck
module.exports.shell = shell
module.exports.comment = comment
